Ported 1.16.1 biomes-api-v1 to 1.16.2

shartte opened the pull request:
Public API changes:
- Removed flagging biomes as suitable for player spawns since that is now handled in the Biome Builder.
- Changed API over to RegistryKey's, not because it's necessary, but because it is more ergonomic since Vanilla Biomes in BuiltInBiomes are now all exposed as keys, rather than instances.

I also increased Gradle-Memory-Settings to 3G since I saw OOM errors in unrelated modules.

This aims to simply restore the biome module for now, and split off the work required to post-process biomes so it becomes more manageable. (See #1036)

This PR in conjunction with #1052 will enable mods to port from 1.16.1 in a safe manner by moving from Registry.* to BuiltInRegistry.* for their WorldGen objects. After this, the next step is to provide simple APIs for adding carvers/features/structures to other biomes (See #1036).

shartte committed:
Ported 1.16.1 biomes-api-v1 to 1.16.2.
Public API changes:
- Removed flagging biomes as suitable for player spawns since that is now handled in the Biome Builder.
- Changed API over to RegistryKey's, not because it's necessary, but because it is more ergonomic since Vanilla Biomes in BuiltInBiomes are now all exposed as keys, rather than instances.

CheaterCodes commented:
This is how I should have started when doing my biome api pr, however there's a few things that we shouldn't port over one to one. Will give this a review tomorrow.

i509VCB commented on fabric-biomes-v1/build.gradle:
```diff
@@ -1,2 +1,2 @@
 archivesBaseName = "fabric-biomes-v1"
-version = getSubprojectVersion(project, "1.0.2")
+version = getSubprojectVersion(project, "1.1.0")
```
This effectively is a new module, so you should probably rename it to `fabric-biome-api-v1`

shartte replied:
Is it? It removes one public method and otherwise changes Biome arguments to RegistryKey<Biome>?
Granted. that _is_ a breaking change but does that justify changing the module?

i509VCB replied:
The entire module is effectively detached from every older version of biome api. Plus these are breaking changes with a module that was never re-released with .2.

The other biomes PR has the consensus by modmuss that naming and packages should be adapted due to an effective full break.

modmuss50 replied:
Its 100% atleast a 2.0.0. As for a new module no. -v2 would only need to be used where the old API was being kept around for backwards compat.

i509VCB commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/AddHillsLayerAccessor.java:
```diff
@@ -0,0 +1,38 @@
+	 * of the sunflower plains, which is derived from plains.
+	 */
+	@Accessor
+	static Int2IntMap getField_26727() {
```
Please name the method, this just confuses me.

shartte replied:
I am afraid to rename it :-|
I'll check out what the consequences of that are w.r.t. accessors.

i509VCB replied:
If you want to rename the getter, specify the field name in the accessor annotation, i.e.:

```java
@Accessor("field_26727")
static Int2IntMap getFoo() {...
```

shartte replied:
Will I have to name it fabric_xxyz in this case or does the accessor guarantee that conflicts are irrelevant still apply here?

i509VCB replied:
Very rarely you'll have a conflict, (key bindings v0 being a former example of this). Should be fine

shartte replied:
Changed to `getBaseToVariantMap`

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeUtils.java:
```diff
@@ -138,7 +148,7 @@ public static void injectBiomesIntoClimate(LayerRandomnessSource random, int[] v
 			// Vanilla biome; look it up from the vanilla array and transform accordingly.
 
-			result.accept(transformBiome(random, BuiltinRegistries.BIOME.get(vanillaArray[(int) reqWeightSum]), climate));
+			result.accept(transformBiome(random, Biomes.fromRawId(vanillaArray[(int) reqWeightSum]), climate));
```
I would suggest using `InternalBiomeUtils.fromRawId` similar to `InternalBiomeUtils.getRawId` so we can make them both use a DynamicRegistryManager if needed

shartte replied:
Vanilla still uses Biomes.fromRawId itself in a place where a DynamicRegistryManager is unavailable, so this won't help.

CheaterCodes replied:
You mean `BiomeLayer.sample`? I would've changed that too.
Dunno, to me that would feel better.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeUtils.java:
```diff
@@ -148,7 +158,22 @@ public static void injectBiomesIntoClimate(LayerRandomnessSource random, int[] v
-	public static void recreateChunkGenerators(GeneratorOptions generatorOptions) {
-		((DimensionOptionsAccessor) (Object) generatorOptions.getDimensionMap().get(DimensionOptions.NETHER)).setChunkGenerator(DimensionTypeAccessor.createNetherGenerator(generatorOptions.getSeed()));
+	public static int getRawId(RegistryKey<Biome> key) {
+		return BuiltinRegistries.BIOME.getRawId(BuiltinRegistries.BIOME.method_31140(key));
```
I would suggest capturing the `biomeRegistry` in `VanillaLayeredBiomeSource.<init>` and use that one theoughout for the conversion `RegistryKey <-> RawId`. That should be compatible with using the DynamicRegistryManager then (and I think also json biomes).
We'd have to check if that capture covers all use cases, but I think it does.

shartte replied:
This will still not fix the Vanilla logic that also accesses the raw-id mapping directly and thus just complicates the logic needlessly.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeUtils.java:
```diff
@@ -148,7 +158,22 @@ public static void injectBiomesIntoClimate(LayerRandomnessSource random, int[] v
+	 * if mods register their biomes only in {@link BuiltinRegistries#BIOME}, and not using the
+	 * private method in {@link Biomes}.
+	 */
+	public static void ensureIdMapping(RegistryKey<Biome> biomeKey) {
```
This validation doesn't seem great to me. How does this interact with registering to `DynamicRegistryManager.BUILTIN` or using json biomes?
I'd suggest dropping this validation for the initial PR.
This seems like it could cause issues, and that's the last thing we want validation to do.

shartte replied:
It doesn't interact with it. This API will be used by modders in their Mod initializers, when no dynamic registry manager exists. The only way to ensure only properly registered Biomes are used is to validate against the BuiltInRegistries, and follow the same general approach Vanilla does.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinAddEdgeBiomesLayer.java:
```diff
@@ -31,18 +32,18 @@
 	@Inject(at = @At("HEAD"), method = "sample", cancellable = true)
 	private void sample(LayerRandomnessSource rand, int north, int east, int south, int west, int center, CallbackInfoReturnable<Integer> info) {
-		Biome centerBiome = BuiltinRegistries.BIOME.get(center);
+		RegistryKey<Biome> centerBiome = Biomes.fromRawId(center);
```
I would suggest using `InternalBiomeUtils.fromRawId` similar to `InternalBiomeUtils.getRawId` so we can make them both use a DynamicRegistryManager if needed.

shartte replied:
Answered above.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinAddHillsLayer.java:
```diff
@@ -48,9 +55,13 @@ private void sample(LayerRandomnessSource rand, LayerSampler biomeSampler, Layer
 		int noiseSample = noiseSampler.sample(chunkX, chunkZ);
 		int processedNoiseSample = (noiseSample - 2) % 29;
-		final Biome biome = BuiltinRegistries.BIOME.get(biomeId);
+		RegistryKey<Biome> key = Biomes.fromRawId(biomeId);
```
I would suggest using `InternalBiomeUtils.fromRawId` similar to `InternalBiomeUtils.getRawId` so we can make them both use a `DynamicRegistryManager` if needed.

shartte replied:
Answered above.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinAddRiversLayer.java:
```diff
@@ -18,42 +18,39 @@
 	private void sample(LayerRandomnessSource rand, LayerSampler landSampler, LayerSampler riverSampler, int x, int z, CallbackInfoReturnable<Integer> info) {
 		int landBiomeId = landSampler.sample(x, z);
-		Biome landBiome = BuiltinRegistries.BIOME.get(landBiomeId);
+		RegistryKey<Biome> landBiomeKey = Biomes.fromRawId(landBiomeId);
```
I would suggest using `InternalBiomeUtils.fromRawId` similar to `InternalBiomeUtils.getRawId` so we can make them both use a `DynamicRegistryManager` if needed.

shartte replied:
Answered above.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinAddRiversLayer.java:
```diff
@@ -18,42 +18,39 @@
-		if (overworldRivers.containsKey(landBiome) && riverBiomeId == RIVER_ID) {
-			Biome riverBiome = overworldRivers.get(landBiome);
-			info.setReturnValue(riverBiome == null ? landBiomeId : BuiltinRegistries.BIOME.getRawId(riverBiome));
+		if (overworldRivers.containsKey(landBiomeKey) && Biomes.fromRawId(riverBiomeId) == BuiltInBiomes.RIVER) {
```
I would suggest... Ok, I think you get the point.

shartte replied:
Answered above.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinSetBaseBiomesLayer.java:
```diff
@@ -96,16 +82,16 @@ private void injectCoolBiomes(LayerRandomnessSource random, int value, CallbackI
 	private void transformVariants(LayerRandomnessSource random, int value, CallbackInfoReturnable<Integer> info) {
 		int biomeId = info.getReturnValueI();
-		Biome biome = BuiltinRegistries.BIOME.get(biomeId);
+		RegistryKey<Biome> biome = Biomes.fromRawId(biomeId);
```
Same thing again.

shartte replied:
Answered above.

CheaterCodes commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinMultiNoiseBiomeSource.java:
```diff
@@ -20,33 +20,45 @@
+ * This Mixin is responsible for adding mod-biomes to the NETHER preset in the MultiNoiseBiomeSource.
+ */
+@Mixin(MultiNoiseBiomeSource.Preset.class)
 public class MixinMultiNoiseBiomeSource {
```
Just gonna mention that I like my approach [in my fork](https://github.com/CheaterCodes/fabric/blob/biome-api/fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MultiNoiseBiomeSourcePresetMixin.java) better, but this is probably fine too.

shartte replied:
That might actually be better. I'll investigate adopting that approach.

shartte replied:
Adopted your approach

CheaterCodes commented on fabric-biomes-v1/src/testmod/java/net/fabricmc/fabric/test/biome/FabricBiomeTest.java:
```diff
@@ -16,26 +16,52 @@
+
+	@Override
+	public void onInitialize() {
+		Registry.register(BuiltinRegistries.BIOME, TEST_CRIMSON_FOREST.getValue(), DefaultBiomeCreator.createCrimsonForest());
```
I know you disagree, but I'm still gonna comment that I don't think we should use `BuiltinRegistries`, but `DynamicRegistryManager.BUILTIN`. We can easily provide a getter for that registry too.

shartte replied:
Paired with #1052 this is the easiest to explain and safest approach of registering worldgen objects, since it is also very easy to follow through the code (and check against Vanilla's behavior)

CheaterCodes commented on fabric-biomes-v1/src/testmod/java/net/fabricmc/fabric/test/biome/FabricBiomeTest.java:
```diff
@@ -16,26 +16,52 @@
-		Biome customPlains = Registry.register(BuiltinRegistries.BIOME, new Identifier(MOD_ID, "custom_plains"), DefaultBiomeCreator.createPlains(null, false));
-		OverworldBiomes.addBiomeVariant(Biomes.PLAINS, customPlains, 1);
+		OverworldBiomes.addContinentalBiome(BuiltInBiomes.END_HIGHLANDS, OverworldClimate.DRY, 0.5);
 	}
```
Would be nice if we could get this to work using json biomes, even if you think we shouldn't use those.
Putting this here cause it'd be part of the test mod.

shartte replied:
I don't think defining the same thing you could define with a constructor call in a JSON file justifies the added complexity for this API and everyone else involved. There is literally nothing else you gain from doing it in a JSON file.

Datapacks don't factor into this. You can _still_ override biome definitions from data packs like you can with Vanilla biomes, and defining a _new_ biome in a datapack ist pointless since the mod that calls this API could not know about it beforehand.

CheaterCodes reviewed (changes requested):
All in all, solid 1 to 1 update of the old api.
I do still think we should move away from BuiltinRegistries.
Even if *you* don't want to use them, I definitely think the API should support `DynamicRegistryManager.BUILTIN` and json Biomes.

CheaterCodes commented:
I also still think that *any* Biome API for 1.16.2+ should be marked as experimental until the underlying vanilla systems become stable. This API may break unrepairably in the next vanilla release.

i509VCB commented on fabric-biomes-v1/src/main/resources/fabric.mod.json:
```diff
@@ -16,8 +16,8 @@
-    "fabricloader": ">=0.4.0",
-    "minecraft": ">=1.16-alpha.20.22.a"
+    "fabricloader": ">=0.9.2",
+    "minecraft": ">=1.16.2"
```
What version did this biome thing occur in?

.2 is probably not optimistic enough about version ranges, but not a major deal tbf

shartte replied:
I did not test this with anything lower than 1.16.2,  hence the lower bound.

modmuss50 replied:
This is fine, no need to go lower. 1.16.1 wont work. No need to care about the snapshots or pre-releases here.

i509VCB commented on fabric-biomes-v1/src/main/resources/fabric-biomes-v1.mixins.json:
```diff
@@ -3,17 +3,16 @@
+    "MultiNoiseBiomeSourceAccessor",
+    "BiomeSourceAccessor",
+    "BiomesAccessor",
+    "AddHillsLayerAccessor"
```
please put these in alphabetical order. This reverse ordering adds a bunch of noise to the pull request

shartte replied:
Done (also: it sucks that the suffix/prefix PR is still in limbo, this is mixed between the two and needs to be changed with that PR)

shartte committed:
Changed to CheaterCode's mixin for modifying the nether biomes, saving us one accessor.

shartte commented:
Rebased onto latest fabric-api and updated the affected yarn mappings. Also adopted CheaterCode's mixin for the nether biomes preset which is more elegant.

modmuss50 reviewed (commented):
This looks good. 2.0.0 should be fine.

Id also say we make this an experimenal API as im sure its going to break. Would also be nice to have an API to modify existing biomes such as adding features to them?

shartte commented:
@modmuss50 I can add it to this if that's desired. Otherwise I was planning to not draw this out too much, since that API will draw more bikeshedding, so I planned a follow-up PR building on this, adding it to this module.
Here's a draft of the public API surface I had in mind: https://gist.github.com/shartte/047f19d191922ab2e96d4957b7bf961f

Same for the Ender-API, really.

Should I make this 2.0.0 and rename the module as @i509VCB requested?

I am not sure what "experimental API"  really means in terms of Minecraft. It's the only thing people can use in 1.16.2 and within that release I'd not consider it experimental :-P

i509VCB commented:
Experimental APIs are described in this RFC: https://github.com/FabricMC/fabric/issues/499

shartte committed:
Move Bonanza

shartte committed:
Finished moving and updating of the module, added experimental markers.

shartte commented:
@modmuss50 @i509VCB I've made the requested changes:

- Marked as experimental by adding `@Deprecated` to the public API and noting in the class-doc why
- Moved the module to fabric-biome-api-v1
- Changed module version to 2.0.0

Due to the rename, the diff has now gone to shit, sadly. I'd rather do any further enhancements of the Biome API in new PRs, after this is merged.

**edit:** May I again note how pointless this experimental mark is? ANY API in this repository is subject to be removed without prior notice due to upstream changes, since we usually don't receive such prior notice from Mojang. This very API was removed without prior notice, and was _not_ marked experimental. So, what exactly is the point? Other than needlessly scaring people away from an API they should be encouraged to use to work with biomes?

shartte committed:
Adjusted experimental notes.

modmuss50 commented:
I would like to get a version of my previously approved PR into this: https://github.com/FabricMC/fabric/pull/982

I can commit this to this branch if you like? It should be fairly easy to get updated I belive it just needs a bit of hardnening to ensure that the feature has been registered. (I have this in TR/RC)

shartte commented:
Player has pointed out that I forgot to rename the package. I'll fix that first. Are you sure you want to add to this branch? The renames make any further changes practically impossible to review.
Alternative: Can you make a separate branch in this repo, and just change the target branch of this PR to that?

shartte committed:
Finished rename from biomes-v1 to biome-api-v1

modmuss50 commented on gradle.properties:
```diff
@@ -1 +1 @@
-org.gradle.jvmargs=-Xmx2560M
```
Can you undo this? This will make it fail to build on jenkins, should be better with 0.5 loom.

modmuss50 commented:
Im happy with this. Can get it in along with #1052

shartte committed:
Reverted memory settings in preparation for loom-0.5

Player3324 reviewed (approved):
LGTM, thanks. Let's get this in and hope it all works :D

MarcusElg commented:
I guess this only is for adding new biomes? You still can't modify existing biome (aka add new features to them)?

oliviathevampire commented:
@MCrafterzz that can be done normally tho, it's just adding features to vanilla biomes we need

MarcusElg commented:
> @MCrafterzz that can be done normally tho, it's just adding features to vanilla biomes we need

Vanilla biomes what was I meant, sorry for confusion
